### ISSUE TITLE
Why does curve_fit not provide a R^2 score?

### ISSUE THREAD

**Issue by jsh9** [scipy/scipy#8439]
Why does curve_fit not provide a R^2 score (i.e., coefficient of determination)?  It is only implemented within linregress.

From Wikipedia (https://en.wikipedia.org/wiki/Coefficient_of_determination), R^2 does not seem to be restricted to only the linear model. In fact, it is implemented in the `fit` function of MATLAB, and also in `sklearn.metrics.r2_score`.

Is it possible to include R^2 in curve_fit in a future release?

### Scipy/Numpy/Python version information:
Python 3.6.3
numpy 1.13.3
scipy 0.19.1

**Comment by pkofod** [comment]
You should not use R-squared to chose between models in non-linear least squares problems. You can always *calculate* it of course, but it will not give you the answer to the question you think you're asking. 

If we use the source you provided (wikipedia) it states that: "In some cases the total sum of squares equals the sum of the two other sums of squares defined above", that is SS_res + SS_reg = SS_tot in their terminology. That is true for OLS (that is, a linear *in parameters* least square problem), but **not** nonlinear least squares. Wiki goes on to say "See Partitioning in the general OLS model for a derivation of this result for one case where the relation holds."

The ultimate problem is that the whole basis for seeing how close R-squared is to 0 or 1 goes away, because R-squared is no longer restricted to be in that interval, and you cannot use R-squared of two different models to assess their relative merits either!

`curve_fit` should not calculate R-squared, as it will very likely cause the uninformed user to draw incorrect conclusions.

**Comment by jsh9** [comment]
With all due respect, I do not need a lecture on the appropriateness of R2 for nonlinear least square.  I already knew the implications of R2 prior to opening this issue 10 months ago.

The focus of my request was on the lack of code functionality, especially because comparable libraries (MATLAB and sklearn) offer it.  Why should a software library hamstring some more informed users because some less informed users would misuse certain functionalities?

Instead, `scipy` can add a paragraph of warning in the documentation, just as in the documentation of its own `f_oneway` function noting the three assumptions for one-way ANOVA: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.f_oneway.html.

**Comment by pkofod** [comment]
> With all due respect, I do not need a lecture on the appropriateness of R2 for nonlinear least square. I already knew the implications of R2 prior to opening this issue 10 months ago.

Okay, that may very well be possible. What you wrote was

> From Wikipedia (https://en.wikipedia.org/wiki/Coefficient_of_determination), R^2 does not seem to be restricted to only the linear model.

I did not read that as if you were fully aware of the shortcomings of R^2 in these types of situations. Even so, I would have written what I wrote to frame my point of view: it should not be implemented.

> The focus of my request was on the lack of code functionality

Yes, and many software packages exclude this functionality _on purpose_!

>  Why should a software library hamstring some more informed users because some less informed users would misuse certain functionalities?

Because many users are not informed, and there's a significant risk it causes more harm than good in this case. Seems like Matlab and scikitlearn decided to go in another direction; can't tell you why. To quote Douglas Bates:

>    There is a good reason that an nls model fit in R does not provide r-squared - r-squared doesn't make sense for a general nls model.
>
>    One way of thinking of r-squared is as a comparison of the residual sum of squares for the fitted model to the residual sum of squares for a trivial model that consists of a constant only. You cannot guarantee that this is a comparison of nested models when dealing with an nls model. If the models aren't nested this comparison is not terribly meaningful.
>
>    So the answer is that you probably don't want to do this in the first place.

I'm not a SciPy developer, but if you really think this is a good idea, maybe provide compelling evidence that this is a feature that is useful to have? A significant part of the statistical software community seems to be against providing the functionality, and it's not like no one knows how to calculate it - they specifically don't want to.

**Comment by jsh9** [comment]
When I said "R^2 does not seem to be restricted to only the linear model" in my original post, I meant that one could calculate R^2 with only "`y_true`" and "`y_predicted`".  How much useful the calculated R^2 was another matter, and was outside the scope of my original post.

Since neither of us is a scipy developer, and since this is more of a design philosophy issue, let's let them decide whether to add this functionality or not.

**Comment by rgommers** [comment]
It's unclear what is better here. `linregress` provides an R value (note R, not R^2). The whole least squares business and consistency between `linregress`, `curve_fit`, `leastsq`, `least_squares`, etc. needs someone to sit down and make a single design review/proposal.

**Comment by jsh9** [comment]
Thanks for revisiting this ticket.

In my opinion, taking a look at how MATLAB handles this is always a good idea:
https://www.mathworks.com/help/matlab/data_analysis/linear-regression.html
https://www.mathworks.com/help/curvefit/fit.html
https://www.mathworks.com/help/curvefit/least-squares-fitting.html
(and some other documents)

By the way, the example of scipy.linregress (https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.linregress.html) that calculates "R square" is very misleading. I would suggest fixing it. The square of that "R value" is not the same as the coefficient of determination (what people usually associate with R-square).

**Comment by rkern** [comment]
FWIW, I consider the omission of R^2 in the nonlinear cases to be a feature. I am -1 on its addition as an automatically-calculated statistic. If you do want to calculate it anyways in places where we don't automatically calculate it, `sklearn.metrics.r2_score` is a fine function to use.

**Comment by jsh9** [comment]
Having to import `sklearn` to do such a simple calculation feels somewhat an overkill. Oftentimes I need to work on computers with only numpy/scipy/matplotlib installed --- no sklearn.
I think implementing a function such as `numpy.r2_score()` could be a sensible middle ground.

**Comment by rkern** [comment]
As it is such a simple calculation, you can also vendorize it in such situations.

**Comment by jsh9** [comment]
I've written a r2_score() function myself and use it where I don't have sklearn. But this is still not a trivial calculation. It would still be great to see it implemented in numpy. After all, `numpy.corrcoef()` is not a difficult calculation either, yet it is implemented in numpy.

**Comment by rkern** [comment]
The difference is primarily one of timing. `numpy` (then `Numeric`) had `corrcoeff` long ago before there were any statistical packages built on top of `Numeric`. Now we do have `sklearn` built on top of `numpy`. All of the functions implemented in `sklearn` are useful. The fact that it would make it convenient for people who want to use them but, for some reason, can't install more than `numpy` and `scipy` is not sufficient reason alone to promote any of them up. At this stage of `numpy`'s and `scipy`'s development, there has to be something more.

`sklearn.metrics.r2_score` makes perfect sense in its current location. `sklearn` has a bunch of different linear regressions and also a bunch of different regression metrics with the expectation that users will mix and match (with the standardized conventions to support that). `r2_score` would be more out of place in `scipy` (and very out of place in `numpy`).

**Comment by jsh9** [comment]
Thanks for this explanation.

**Comment by mdhaber** [comment]
@steppi I thought it would be good to get your opinion on this one.

**Comment by steppi** [comment]
> @steppi I thought it would be good to get your opinion on this one.

I agree that $R^2$ is a flawed metric for nonlinear regression. Including it in the output seems like it’s condoning and encouraging it’s use in this context. -1 for its inclusion in the output of `curve_fit`.

**Comment by andyfaff** [comment]
I think this is going to be  a wontfix

**Comment by andyfaff** [comment]
There's no consensus to add, if anything it's the other way.

**Comment by mdhaber** [comment]
Thanks. That was my conclusion, too.
